### PR TITLE
Mangle exported symbols 

### PR DESCRIPTION
**Pull request by mjbvz**
Porting https://github.com/microsoft/vscode/pull/180462 to a branch in our main repo

For https://github.com/microsoft/vscode/issues/180461. This prototype tries to mangle exported functions, constants, and classes

Current total savings: 1910kb
New total savings: 4077kb

### CODE REVIEW

File: src/vs/server/node/server.main.ts [hunk 1]
```diff
@@ -58,13 +58,17 @@ args['extensions-dir'] = args['extensions-dir'] || join(REMOTE_DATA_FOLDER, 'ext
 
 /**
  * invoked by server-main.js
+ *
+ * @skipMangle
```
[the remark this answers is not in the document]
Comment (mjbvz) [reply src/vs/server/node/server.main.ts:62]: We need to skip mangling a handful of symbols that are used in plain old js files that don't belong to the same TSConfig as the rest of our sources

In this case, I can instead add this file to our list of entry points but there are some non entry point files that also get used in this way. Overall though I only had to add `@skipMangle` around 15 times to our codebase so it shouldn't be a huge issue

**Comment by mjbvz** [comment]
Finally got a successful full build: https://dev.azure.com/monacotools/Monaco/_build/results?buildId=215714 Main issue was that I was accidentally mangling ambient type declarations, which messed up some referenced libraries in remote scenarios (this only happens in full builds

Local testing all seems ok although I haven't run though every scenario. Will continue testing and hopefully get this merged in June

**Comment by jrieken** [comment]
@mjbvz Can you share numbers how much slower the (product) build will be because of this?

**Comment by mjbvz** [comment]
@jrieken This makes the `compile` set of the build about 10 minutes slower: going from 13 minutes to 22 minutes. The other builds steps are not effected

This slow down is due to the large number of `rename` requests we make against typescript. I brought this up to the TS team but they seem unlikely to address it (the rational being that renames are typically triggered only by occasional UI actions). I also use worker threads to improve performance, but can only use two workers at the moment to avoid memory issues on the CI machines

**Comment by jrieken** [comment]
fyi @alexdima - you were wondering about the additional costs of this

**Review by jrieken — APPROVED** [review]

**Pull request merged**